docs/adr/adr-013-blocksync-overhaul-part-2.md

```diff
@@ -0,0 +1,421 @@
+
+### Key Design Decisions
+
+- __Use [go-libp2p][go-libp2p]'s software stack to build the protocols.__ It's extensively used in `celestia-node` already,
+proven to be robust and provides all the requirRed components and tools for the new protocols.
+
+- __Use [protobuf][proto] schema for protocol(TODO Link) message (de)serialization.__ Protobuf is a common (de)serialization
+standard which is already used in celestia-node. Even thought there are better alternatives, their benefits are
+negligible comparing with continuing Protobuf usage for consistency and simplicity.
+
+- __FN/LNs keep existing EDS retrieval logic(TODO Link?).__ `celestia-node` already provides means to get EDS from
+the network. It is slower than block production for the EDS sizes __>128__, but has higher retrieval guarantees during
+data withholding attack. (link)
+
+- __FN/LN/BNs leverage implemented discovery mechanism outlined in ADR-008(TODO Link).__ The discovery in a node maintains connections
+with FN-subnetwork. This ensures the node will receive new notifications from `ShrEx/Sub` and can request data via
+`ShrEx/EDS` and `ShrEx/ND`.
+
+- __FN/BNs notify connected LN/FNs when they are ready to serve a recent EDS using via [`ShrEx/Sub`](#ShrEx/Sub Protocol).__
+This lazy-push approach provides reliable EDS dissemination properties and turns out to be simpler than technics similar
+to Bitswap's wantlist.(TODO link)
+
+## Protocols Specs
+
+### ShrEx/EDS Protocol
+
+ShrEx/EDS is a pull-based protocol with client-server model, where LNs are clients, BNs are servers
+and FNs are both. The protocol has only one request-response interaction followed by ODS stream, which is then
+reconstructed into EDS
+
+The protocol is designed over plain libp2p streams with `/shrex/eds/0.0.1` as the protocol ID.
+
+#### Protobuf Schema
+
+##### Request
+
+```protobuf
+syntax = "proto3";
+
+message EDSRequest {
+  bytes hash = 1; // identifies the requested EDS.
+}
+```
+
+##### Response
+
+```protobuf
+syntax = "proto3";
+
+enum Status {
+  OK = 100; // data found
+  NOT_FOUND = 200; // data not found
+  REFUSED = 201; // request refused
+}
+
+message EDSResponse {
+  Status status = 1;
+}
+```
+
+#### Streaming
+
+After flushing `EDSResponse`, server starts streaming ODS file to client over the libp2p stream.
+
+#### Backpressure
+
+The streaming part of the protocol does not provide any mechanics for backpressure and offloads them to layer below.
+Client-side buffer with reasonable default size may still be added.
+
+[Backpressure][backpressure-quic] is inherited from the QUIC. The underlying default reliable transport and stream
+multiplexing protocol used in `celestia-node`.
+
+The common TCP transport and yamux multiplexer combination, does not provide reliable backpressure on stream level,
+unlike QUIC. Therefore, using TCP+yamux is not recommended and may lead to unexpected stream and connections resets in case of highly
+congested links.
```

Comment:
Q: Does the streaming only work relaibly using QUIC?

Reply:
I think so